Adding regularization objective option to parmest

sscini opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
Currently, the only default objective is the standard SSE objective. This edit provides the capability to add a 
regularization term to the SSE objective with a prior FIM and reference parameter values.


## Changes proposed in this PR:
- Added prior_FIM and theta_ref as keyword arguments for Estimator
- Added options within Estimator to apply the regularization term to SSE if the new arguments were defined 

## TODO before converting from draft

- [x] Receive feedback from collaborators on logical setup
- [x] Finish edits and debug
- [x] Confirm functionality with examples

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

sscini committed:
Add regularize term function, and modified SSE to add option

sscini committed:
Fixed term in function

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Update parmest.py

sscini committed:
Revert "Update parmest.py"

This reverts commit 055f14cbdde28cc4c45fa3ec42a9640053f3ddd1.

sscini committed:
Reapply "Update parmest.py"

This reverts commit 626c1b9ba76a2f46e1304226130ebd7a50aad51b.

sscini committed:
Add theta_ref and prior_FIM to keyword args

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Adjust naming convention to ensure consistency.

sscini commented:
@djlaky @adowling2 Please provide early feedback

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -234,6 +234,21 @@ def SSE(model):
+
+    Added to SSE objective function
+    """
+    expr = ((theta - theta_ref).transpose() * prior_FIM * (theta - theta_ref) for theta in model.unknown_parameters.items())
```
Does this run? My intuition is that you need to write out the matrix multiplication and cannot use matrix multiplication.

adowling2 reviewed (commented):
@sscini See my early feedback.

sscini committed:
Update parmest.py, DoE meeting work

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini commented:
UPDATE (07/15/25); Intended to be added AFTER Shammah's weighted SSE PR and follow a similar format, adding another option for available objectives using Enums.

Currently preparing for next stage of review

sscini committed:
Updated for next round of review

sscini commented:
@adowling2 @djlaky Made some edits to this finally. Meant to follow Shammah's layout in PR #3535, waiting to merge, currently some conflicts I cannot resolve on my end.

Made it's own full objective instead of a term to add. Also added weighting term.

Ready for next round of review.

sscini committed:
Added regularization weight to exp class initialize

sscini committed:
Still debugging, progress as of 7/15/2025

sscini commented:
Will not be actively worked on before finalization of _Q_opt_blocks. Closing for now to resume in the future.

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Update parmest.py

sscini committed:
Fixed issues with merging changes

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
L2 regularization working

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini commented:
Got L2 regularization working, still need to modify to improve implementation slightly. Made prior_FIM and theta_ref as pd.DataFrames because this is how parmest would return them, and would not be reliant on assuming order is correct.

sscini committed:
Made the theta_ref a series.

sscini commented:
Made theta_ref a pd.Series so if you take theta and invert the covariance outputs from theta_est(), you can feed them directly back in.

sscini committed:
Updated doc strings.

sscini committed:
Update parmest.py

sscini committed:
added helper comments.

sscini committed:
Updated implementation to add vectorized L2 calc, and reviewer questions

sscini committed:
Adjusted logging, ran black

sscini committed:
Changed value to avoid zero for L2 term

sscini committed:
Commented out or removed print statements

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Chose to apply in compute_covariance

sscini committed:
Update parmest.py

Weight was not symbolic, so was not affecting parameter estimation. Made significant changes to adjust structure so the regularization weight is calculated within the Estimator class, and applied to the covariance update as well.

sscini committed:
Made weight default 1

sscini committed:
Ran black

sscini committed:
Adjusted doc strings, ran black

sscini committed:
Update regularization_example.py

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Added validation for prior_FIM and fixed inconsistencies

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini commented:
Thursday night (03/05/26): Made changes to architecture based on additional recommendations from ND Pyomo Dev team, and added a separate testing file for tests I am considering adding to review.

All feedback is appreciated. Main items of interest in review:

1. User interface for the new items needed in the Estimator class: add to Estimator or theta_est? (chose estimator), make separate inputs (current) or have arguments passed in as a dictionary/class?
2. Is regularization weight input useful for users to be able to adjust results? Or keep weighting strictly to prior_FIM?
3. Should users be able to pass in their own regularization scheme? Currently no
4. Pass in prior_FIM and theta0, currently as pd.DataFrame and pd.Series. Support additional?
5. Review new test additions in separate file

Thanks!

sscini committed:
Added L1 reg support working prototype, black

sscini committed:
Fixed testing issue, and example issues

adowling2 reviewed (commented):
There are a few good questions within this PR to discuss with the Pyomo team. I think this PR depends on https://github.com/Pyomo/pyomo/pull/3789

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Added tests to correct files, ran black.

sscini committed:
Updated model structure, reordered, ran black

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Archived L1 for now, removed from here, ran black

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1610,6 +2212,8 @@ def _get_sample_list(self, samplesize, num_samples, replacement=True):
         return samplelist
 
+    # @Reviewers: Currently regularization chosen with objective at Estimator initialization,
+    # Would it be preferable to have regularization choice as an argument in the theta_est function instead?
```
Implementation for L2 in this PR still on Estimator class, makes sense to go with the objective selection in my mind. Can also revisit with weighted smooth L1 implementation in future.

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Delete pyomo/contrib/parmest/scenarios.csv

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

blnicho commented on pyomo/contrib/parmest/tests/test_examples.py:
```diff
@@ -54,6 +55,19 @@ def test_likelihood_ratio_example(self):
+        l2_obj, l2_theta, _ = results["L2"]
+
+        self.assertTrue(math.isfinite(float(l2_obj)))
+        self.assertTrue(math.isfinite(float(l2_theta["rate_constant"])))
```
Why are you only verifying that the outputs are finite and not checking for specific values?

sscini committed:
Fixed numerical issue, ran black

sscini committed:
Merge branch 'main' into parmest_obj_regularization

blnicho committed:
NFC: Fix formatting

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -149,6 +149,183 @@ def SSE_weighted(model):
+
+        return 0.0  # No regularization if no common parameters
+
+    elif len(common_params) < len(set(prior_FIM.columns).union(set(prior_FIM.index))):
```
```suggestion
    elif len(common_params) < len(set(prior_FIM.columns)):
```
When you validate the `prior_FIM` you confirm that `prior_FIM.columns` and `prior_FIM.index` are the same so I don't think you need to take the union here.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -149,6 +149,183 @@ def SSE_weighted(model):
+    common_params = [
+        p
+        for p in current_param_names
+        if p in prior_FIM.columns and p in prior_FIM.index
```
```suggestion
        if p in prior_FIM.columns
```
You have already validated that `prior_FIM.columns` and `prior_FIM.index` are the same so do you need to check for containment in both?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -149,6 +149,183 @@ def SSE_weighted(model):
+    else:
+
+        logger.info(
+            "All parameters in the Prior FIM match the Model parameters. Regularization will be applied to all parameters."
```
Could you break this into multiple lines to stay within the 88-column limit?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -149,6 +149,183 @@ def SSE_weighted(model):
+
+    # For the reference theta, we also subset to the common parameters. If theta_ref is None, we create a zero vector of the right size.
+    if theta_ref is not None:
+        if not isinstance(theta_ref, pd.Series):
```
Could you break this into multiple lines to stay within the 88-column limit?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -149,6 +149,183 @@ def SSE_weighted(model):
+    #             qij = float(sub_FIM.loc[i, j])
+    #             if qij != 0.0:
+    #                 dj = name_to_var[j] - float(theta0.loc[j])
+    #                 l2_expr += qij * di * dj
```
I tested this locally and I think the vectorized form should work fine here. The only thing to be careful of is to not take the value of any Vars you want preserved in the final expression (which you're following). I think this comment can be removed.

mrmundt commented on pyomo/contrib/parmest/examples/rooney_biegler/regularization_example.py:
```diff
@@ -0,0 +1,92 @@
+    # relative_error = abs(theta['asymptote'] - 19.1426) / 19.1426
+    # assert relative_error < 0.01
+    # relative_error = abs(theta['rate_constant'] - 0.5311) / 0.5311
+    # assert relative_error < 0.01
```
Commented out intentionally or accidentally?

sscini replied:
Accidental. Was left commented out after trying other values to debug something. Resolving now

sscini commented:
@blnicho Thank you for fixing the formatting issue. I tried to resolve the merge conflict after _Q_opt and was not near my laptop. Both of the failing tests are not related to my tests.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -353,8 +530,12 @@ class ObjectiveType(Enum):
 
 
+class RegularizationType(Enum):
+    L2 = "L2"
```
Is the plan to eventually add more regularization types? If so, this is fine; if not, why bother with this?

sscini replied:
Thank you for your comments! Yes this PR a week ago actually included a version of L1 regularization too, but we were not confident in the smooth approximation implementation and wanted to simplify for the review process here. Smooth L1 will be a future PR, and we plan to add more based on different parameter priors.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -513,11 +713,36 @@ def compute_covariance_matrix(
+
+        if obj_function == ObjectiveType.SSE_weighted.value:
+            expanded_prior_FIM *= (
+                0.5  # apply the 0.5 convention for weighted SSE regularization
```
I don't know enough about weighted SSE regularization, so this may be a stupid question - but is 0.5 as a convention cite-able in any way? Is there a reason why the convention is 0.5 that can be verified?

(Basically, is this a magic number? And if so, why THIS magic number?)

sscini replied:
This was poorly explained here, thank you for comment. Will cite this more in the code comments and include in PR discussion. @slilonfe5 please confirm. When implementing SSE and SSE weighted from the log likelihood, the SSE_weighted included a 1/2 in front of the term to make the derivatives cleaner. The SSE did not. So multiplying the regularization term by 0.5 for SSE weighted keeps the scaling the same.

slilonfe5 replied:
@mrmundt Stephen is right. Stephen added the 0.5 to the regularization term of the SSE_weighted objective because the SSE_weighted objective has a 1/2 in it, which comes naturally from the log-likelihood function.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -788,10 +1083,67 @@ def __init__(
 
+        if isinstance(regularization, str):
+            try:
+                self.regularization = RegularizationType(regularization)
```
Totally agree with this paradigm - but again, are there any plans to add different types of regularization? Otherwise, it's abstraction for no good reason.

sscini replied:
Plan is to expand to include multiple, but implementation details get more complex if the regularized term is not twice differentiable.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -788,10 +1083,67 @@ def __init__(
+            self.prior_FIM = None
+            self.theta_ref = None
+            self.regularization_weight = None
+            self.regularization_epsilon = None
```
What is `regularization_epsilon`? It isn't described in the docstring and it doesn't look like it's being used anywhere.

sscini replied:
At one point this PR had L1 and L2 regularization, but was simplified to L2. A regularization epsilon was used for a smooth approximation of L1, since the absolute value would not be twice differentiable. I tried to remove all of them but missed this one. Removing now.

blnicho reviewed (commented):
@sscini I have a few minor comments and questions but otherwise this looks fine.

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Addressing review comments, in progress. Ran black

sscini committed:
Addressed changes, fixed bug, ran black

sscini committed:
Removed unneeded import

sscini committed:
Fixed indexed param inconsistency, ran black

sscini commented:
@blnicho @mrmundt Thank you both for all your comments. Running new checks now, but I believe all are addressed.

slilonfe5 commented:
@sscini @blnicho I’ve noticed that the code coverage of ParmEst is now at 81%, down from 90%. Is this something to fix in this PR while we’re at it? So someone in the future doesn’t have to worry about this.

sscini commented:
> @sscini @blnicho I’ve noticed that the code coverage of ParmEst is now at 81%, down from 90%. Is this something to fix in this PR while we’re at it? So someone in the future doesn’t have to worry about this.

Great point @slilonfe5. I am planning to add a few today. Also, is ParmEst at 81% or my change lines in this code? I think what it says here is my changes are at 81% but I could be wrong.

sscini committed:
Merge branch 'Pyomo:main' into parmest_obj_regularization

blnicho commented:
@sscini @slilonfe5 the 81% patch coverage means that 81% of the lines touched in this PR are covered by the tests. The overall project coverage for the entire repo will remain almost constant at ~90%. 81% patch coverage is a little low so it would be great if we could add a few more tests before merging. If you click on the coverage report it should tell you which lines are uncovered.

sscini commented:
@blnicho Thanks for clarifying! I started working on tests but still need to finalize. Away from computer but should be online have them in by early afternoon in MT.

sscini commented:
@slilonfe5 @adowling2 @blnicho One small design question I wanted to clarify. In this implementation it was assuming the theta and cov were returned as pd.Series and pd.DataFrames, respectively. We changed in _Q_opt to where theta is now a dictionary. 

Should this support: a) current pd Series and DataFrames, b) Dictionary and DataFrames, c) Both dictionaries, or d) both dictionaries and pd objects for either? 

The other option is to convert the output back to pd.Series for theta, but converting dict to a series can be shown in an example if that is the stronger long term design choice. 

All of this can also be decided in a future PR if this is not a large issue.

slilonfe5 commented:
> @slilonfe5 @adowling2 @blnicho One small design question I wanted to clarify. In this implementation it was assuming the theta and cov were returned as pd.Series and pd.DataFrames, respectively. We changed in _Q_opt to where theta is now a dictionary.
> 
> Should this support: a) current pd Series and DataFrames, b) Dictionary and DataFrames, c) Both dictionaries, or d) both dictionaries and pd objects for either?
> 
> The other option is to convert the output back to pd.Series for theta, but converting dict to a series can be shown in an example if that is the stronger long term design choice.
> 
> All of this can also be decided in a future PR if this is not a large issue.

@sscini, I think the design decision (dictionary for theta and pd.DataFrame for cov) from the merged Q_opt PR is better. These should already reflect on this branch after updating it with the Pyomo main.

blnicho commented:
I agree with @slilonfe5, let's stick with the design from the Q_opt PR for now.

sscini commented:
@blnicho With the reviewer approval, when I make new changes will the coverage test rerun each time, or once I am finished do I request the new coverage test?

blnicho commented:
The coverage will be updated every time you make changes to this PR

sscini committed:
Updating reg to use dict, in progress

sscini committed:
Merge branch 'main' into parmest_obj_regularization

sscini committed:
Added tests to improve coverage, ran black

slilonfe5 committed:
Added logger warning to the covariance matrix

slilonfe5 committed:
Merge branch 'main' into parmest_obj_regularization

sscini commented:
@blnicho @mrmundt Please review final version when available. There was one more small change required if you could please add the pre-test inspected label once more.  Otherwise all tests passing. Thanks!

slilonfe5 commented:
> @blnicho @mrmundt Please review final version when available. There was one more small change required if you could please add the pre-test inspected label once more. Otherwise all tests passing. Thanks!

@blnicho @mrmundt Most of the code is now covered. I think this is ready to be merged

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1527,6 +1929,8 @@ def _get_sample_list(self, samplesize, num_samples, replacement=True):
         return samplelist
 
+    # @Reviewers: Currently regularization chosen with objective at Estimator initialization,
+    # Would it be preferable to have regularization choice as an argument in the theta_est function instead?
```
Is this resolved appropriately for this PR? (I don't have an answer to the question; just want to ensure it isn't forgotten if it's still necessary to get an answer.)

sscini replied:
Yes it is resolved. It makes more sense (in current interface) to have the regularization option with the objective option in the estimator. If strong opinion, my next planned PR to add L1 regularization can address. 

Thank you for catching comment. I will remove now

sscini committed:
Final formatting pass, ran black
